collect/collect.go

```diff
@@ -417,58 +417,72 @@ func (i *InMemCollector) collect() {
 				span.End()
 				return
 			}
-			i.processSpan(ctx, sp, "peer")
-		default:
-			select {
-			case msg, ok := <-i.dropDecisionMessages:
-				if !ok {
-					// channel's been closed; we should shut down.
-					span.End()
-					return
-				}
-				i.processTraceDecisions(msg, dropDecision)
-			case msg, ok := <-i.keptDecisionMessages:
-				if !ok {
-					// channel's been closed; we should shut down.
-					span.End()
-					return
-				}
-				i.processTraceDecisions(msg, keptDecision)
-			case <-ticker.C:
-				i.sendExpiredTracesInCache(ctx, i.Clock.Now())
-				i.checkAlloc(ctx)
-
-				// maybe only do this if in test mode?
-				// Briefly unlock the cache, to allow test access.
-				_, goSchedSpan := otelutil.StartSpan(ctx, i.Tracer, "Gosched")
-				i.mutex.Unlock()
-				runtime.Gosched()
-				i.mutex.Lock()
-				goSchedSpan.End()
-			case sp, ok := <-i.incoming:
-				if !ok {
-					// channel's been closed; we should shut down.
-					span.End()
-					return
-				}
-				i.processSpan(ctx, sp, "incoming")
-			case sp, ok := <-i.fromPeer:
-				if !ok {
-					// channel's been closed; we should shut down.
-					span.End()
-					return
-				}
-				i.processSpan(ctx, sp, "peer")
-			case <-i.reload:
-				i.reloadConfigs()
+			drainSpanQueue(ctx, sp, i.fromPeer, "peer", i.processSpan)
+		case sp, ok := <-i.incoming:
+			if !ok {
+				// channel's been closed; we should shut down.
+				span.End()
+				return
 			}
+			drainSpanQueue(ctx, sp, i.incoming, "incoming", i.processSpan)
+		case msg, ok := <-i.dropDecisionMessages:
+			if !ok {
+				// channel's been closed; we should shut down.
+				span.End()
+				return
+			}
+			i.processTraceDecisions(msg, dropDecision)
+		case msg, ok := <-i.keptDecisionMessages:
+			if !ok {
+				// channel's been closed; we should shut down.
+				span.End()
+				return
+			}
+			i.processTraceDecisions(msg, keptDecision)
+		case <-ticker.C:
+			i.sendExpiredTracesInCache(ctx, i.Clock.Now())
+			i.checkAlloc(ctx)
+
+			// maybe only do this if in test mode?
+			// Briefly unlock the cache, to allow test access.
+			_, goSchedSpan := otelutil.StartSpan(ctx, i.Tracer, "Gosched")
+			i.mutex.Unlock()
+			runtime.Gosched()
+			i.mutex.Lock()
+			goSchedSpan.End()
+		case <-i.reload:
+			i.reloadConfigs()
 		}
 
 		i.Metrics.Histogram("collector_collect_loop_duration_ms", float64(time.Now().Sub(startTime).Milliseconds()))
 		span.End()
 	}
 }
 
+func drainSpanQueue(ctx context.Context, span *types.Span, ch <-chan *types.Span, queueName string, processSpanFunc func(context.Context, *types.Span, string)) {
+	// process the original span
+	processSpanFunc(ctx, span, queueName)
+
+	// let't try to process as many spans as we can in the next 100ms
+	// TODO: make timer configurable?
+	timer := time.NewTimer(time.Millisecond * 100)
```

Comment:
Right now each collect loop iteration takes between 10ms to 20ms. Setting the timer to 100ms means we probably are going to process 5-10 items off of the queue each time. Is that enough?
I wonder if we should drain the queue based on the number of items process instead of a timer to make sure we are always draining X amount of items for each iteration.

I don't know which strategy is better. We can try the timer based first to see how it goes

Reply:
I think the timer approach is worth trying. If we want it to be more flexible, we could add a temporary config option while we're testing it out.

Reply:
Yes, maybe we can add a metric here to see how many items we drain from the queue each time?